bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/cache/ExpiringCacheAsync.java

```diff
@@ -0,0 +1,137 @@
+ * issue a fetch and notify callback implementors asynchronously.
+ * This class is not thread safe and requires external looking.
+ *
+ * @author David Graeff - Initial contribution
+ *
+ * @param <V> the type of the cached value
+ */
+public class ExpiringCacheAsync<V> {
+    // The fields are package private on purpose, to allow easier unit testing.
+    final long expiry;
+    ExpiringCacheUpdate cacheUpdater;
+    long expiresAt = 0;
+    boolean refreshRequested = false;
+    V value;
+    final List<Consumer<V>> waitingCacheCallbacks = new LinkedList<>();
+
+    /**
+     * Implement the requestCacheUpdate method which will be called when the cache
+     * needs an updated value. Call {@link setValue} to update the cached value.
+     */
+    public static interface ExpiringCacheUpdate {
+        void requestCacheUpdate();
+    }
+
+    /**
+     * Create a new instance.
+     *
+     * @param expiry the duration in milliseconds for how long the value stays valid. Must be greater than 0.
+     * @param cacheUpdater The cache will use this callback if a new value is needed. Must not be null.
+     * @throws IllegalArgumentException For an expire value <=0 or a null cacheUpdater.
+     */
+    public ExpiringCacheAsync(long expiry, ExpiringCacheUpdate cacheUpdater) throws IllegalArgumentException {
+        this.expiry = TimeUnit.MILLISECONDS.toNanos(expiry);
+        this.cacheUpdater = cacheUpdater;
+        if (expiry <= 0) {
+            throw new IllegalArgumentException("Cache expire time must be greater than 0");
+        }
+        if (cacheUpdater == null) {
+            throw new IllegalArgumentException("A cache updater is necessary");
+        }
+    }
+
+    /**
+     * Returns the value - possibly from the cache, if it is still valid.
+     *
+     * @return the value
+     */
+    public void getValue(Consumer<V> callback) {
+        if (isExpired()) {
+            refreshValue(callback);
+        } else {
+            callback.accept(value);
+        }
+    }
+
+    /**
+     * Invalidates the value in the cache.
+     */
+    public void invalidateValue() {
+        expiresAt = 0;
+    }
+
+    /**
+     * Updates the cached value with the given one.
+     *
+     * @param newValue The new value. All listeners, registered by getValueAsync() and refreshValue(), will be notified
+     *            of the new value.
+     */
+    public void setValue(V newValue) {
+        refreshRequested = false;
+        value = newValue;
+        expiresAt = getCurrentNanoTime() + expiry;
+        // Inform all callback handlers of the new value and clear the list
+        for (Consumer<V> callback : waitingCacheCallbacks) {
+            callback.accept(value);
+        }
+        waitingCacheCallbacks.clear();
+    }
+
+    /**
+     * Returns an arbitrary time reference in nanoseconds.
+     * This is used for the cache to determine if a value has expired.
+     */
+    public long getCurrentNanoTime() {
```

Comment:
Because of tests :) This way I can intercept the call and inject a time and don't need to "wait" for expire tests.